### PR TITLE
`Development`: Refactor the relation between learning goals and exercises

### PR DESCRIPTION
**Pull request by ge65cer**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.

### Motivation and Context

Instructors can create _lecture units_ for a lecture, where they may also choose to add existing exercises as a lecture unit (so called "exercise units").
This is a special case, because title and release date for an _exercise unit_ are taken from the connected exercise.
Instructors may also connect lecture units, including exercise units, to one or more learning goals (this relation is saved in the “learning_goal_lecture_unit” table).

However, when a course does not use lecture units, there exists no way to connect _exercises_ to _learning goals_.

### Description

In this PR, we refactor the relation between _learning goals_ and _exercises_ to **eventually** allow instructors to connect _exercises_ to _learning goals_.
The database table and class attributes already existed (“learning_goal_exercise” table), but were previously not used.
The migration included in this PR moves existing relations between _exercise units_ and _learning goals_ to this table and removes them from the “learning_goal_lecture_unit” table.
**Now, the learning goals for an exercise unit are taken from the connected exercise.**
You should not notice any changes in the client/as an instructor. Client changes for learning goals are coming up in a future PR, so a user interface to connect exercises and learning goals is not provided, yet.

### Steps for Testing

Prerequisites:
- 1 Instructor

1. Log in to Artemis
2. Navigate to Course Administration
3. Go to Learning Goals
4. Ensure you can still create and delete learning goals
5. Make sure you can edit learning goals and connect them to all kids of lecture units (especially exercise units)
6. Also test creating, editing, deleting lectures, exercises, and lecture units

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Instructions | Branches |
|------------|-------:|-----:|
| LearningGoalService.java | 95% | 62% |
| LectureService.java | 95% | 91% |
| LectureUnitService.java | 90% | 83% |
| LearningGoalResource.java | 92% | 63% |
| LectureResource.java | 97% | 78% |

### CODE REVIEW

File: src/main/resources/config/liquibase/changelog/20220903100502_changelog.xml [hunk 1]
```diff
@@ -0,0 +1,21 @@
+<?xml version="1.1" encoding="UTF-8" standalone="no"?>
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog" xmlns:ext="http://www.liquibase.org/xml/ns/dbchangelog-ext" xmlns:pro="http://www.liquibase.org/xml/ns/pro" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog-ext http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-ext.xsd http://www.liquibase.org/xml/ns/pro http://www.liquibase.org/xml/ns/pro/liquibase-pro-3.8.xsd http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.8.xsd">
+    <changeSet author="ge65cer" id="20220903100502-1">
+        <comment>Fill the learning_goal_exercise table with relations currently stored in the learning_goal_lecture_unit table.</comment>
+        <sql>
+            INSERT INTO learning_goal_exercise (learning_goal_id, exercise_id)
+            SELECT lg.learning_goal_id, lu.exercise_id FROM learning_goal_lecture_unit lg
+            LEFT JOIN lecture_unit lu ON lu.id = lg.lecture_unit_id
```
Comment (stefanwaldhauser) [inline src/main/resources/config/liquibase/changelog/20220903100502_changelog.xml:8]: ❓Question: Not a blocker but just out of interest why a left join is used here. There should always be a lecture unit no?

File: src/main/resources/config/liquibase/changelog/20220903100502_changelog.xml [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+            WHERE lu.discriminator = "E";
+        </sql>
+    </changeSet>
+    <changeSet author="ge65cer" id="20220903100502-2">
+        <comment>Remove all relations between exercise units and learning goals from the learning_goal_lecture_unit table.</comment>
```
Comment (stefanwaldhauser) [inline src/main/resources/config/liquibase/changelog/20220903100502_changelog.xml:13]: 💡Idea: Just to be extra safe here, you could have a precondition that the above changeSet has been really executed and we do not lose any data. I know the changeSets should be executed in order, but maybe just to be safe here :) 

See: https://docs.liquibase.com/concepts/changelogs/preconditions.html#:~:text=a%20given%20property.-,changeSetExecuted,-Defines%20if%20the

File: src/main/java/de/tum/in/www1/artemis/domain/LearningGoal.java [hunk 3]
```diff
@@ -115,6 +134,15 @@ public void setConsecutiveCourses(Set<Course> consecutiveCourses) {
         this.consecutiveCourses = consecutiveCourses;
     }
 
+    /**
+     * Ensure that exercise units are connected to learning goals through the corresponding exercise
+     */
+    @PrePersist
+    @PreUpdate
+    public void prePersistOrUpdate() {
+        this.lectureUnits.removeIf(lectureUnit -> lectureUnit instanceof ExerciseUnit);
```
Comment (stefanwaldhauser) [inline src/main/java/de/tum/in/www1/artemis/domain/LearningGoal.java:143]: 😍 I like the approach of using a listener here to guarantee this!

**Review by stefanwaldhauser — CHANGES_REQUESTED** [review]
❌ Good job but I am getting an intern server error when creating a new learning goal connected to exercise units.

File: src/main/resources/config/liquibase/changelog/20220903100502_changelog.xml (hunk 2 above)
Comment (ge65cer) [reply src/main/resources/config/liquibase/changelog/20220903100502_changelog.xml:13]: I [had a pre-condition](https://github.com/ls1intum/Artemis/pull/5642/commits/0234f1d9cfe23859ad8c619ff101c747cead11e2) that the table already exists, but then the Cypress tests were failing to build.

File: src/main/resources/config/liquibase/changelog/20220903100502_changelog.xml (hunk 1 above)
Comment (ge65cer) [reply src/main/resources/config/liquibase/changelog/20220903100502_changelog.xml:8]: Yes, due to our constraints there will always be a lecture unit for the id in the relations table. To be 100% percent safe, I can change this to an inner join.

**Review by stefanwaldhauser — CHANGES_REQUESTED** [review]
I tested it again and the creation of learning goals with exercise units works now and the score is calculated. But when I click on the learning goal the modal says that it is not connected to any lecture units (see screenshot). I guess this is not intended?
image

File: src/main/java/de/tum/in/www1/artemis/domain/lecture/ExerciseUnit.java [hunk 4]
```diff
@@ -62,4 +59,25 @@ public ZonedDateTime getReleaseDate() {
     public void setReleaseDate(ZonedDateTime releaseDate) {
         // Should be set in associated exercise
     }
+
+    @Override
+    public Set<LearningGoal> getLearningGoals() {
+        return exercise == null || !Hibernate.isPropertyInitialized(exercise, "learningGoals") ? new HashSet<>() : exercise.getLearningGoals();
```
Comment (sleiss) [inline src/main/java/de/tum/in/www1/artemis/domain/lecture/ExerciseUnit.java:65]: Is there a known case where the property `learningGoals``is not initialized or is this just as precaution?

File: src/main/java/de/tum/in/www1/artemis/web/rest/LearningGoalResource.java [hunk 5]
```diff
@@ -197,31 +197,30 @@ public ResponseEntity<LearningGoal> updateLearningGoal(@PathVariable Long course
         if (learningGoal.getId() == null) {
             throw new BadRequestException();
         }
-        var learningGoalFromDb = this.learningGoalRepository.findByIdWithLectureUnitsBidirectionalElseThrow(learningGoal.getId());
-        if (learningGoalFromDb.getCourse() == null || !learningGoalFromDb.getCourse().getId().equals(courseId)) {
+        var existingLearningGoal = this.learningGoalRepository.findById(learningGoal.getId()).orElseThrow();
```
Comment (sleiss) [inline src/main/java/de/tum/in/www1/artemis/web/rest/LearningGoalResource.java:200]: Is there a reason you are not using `learningGoalRepository.findByIdElseThrow` here?

File: src/main/java/de/tum/in/www1/artemis/domain/lecture/ExerciseUnit.java (hunk 4 above)
Comment (ge65cer) [reply src/main/java/de/tum/in/www1/artemis/domain/lecture/ExerciseUnit.java:65]: Yes, “learningGoals” in the exercise is a lazy collection. So there might be cases where they are not initialized.

Comment (sleiss) [reply src/main/java/de/tum/in/www1/artemis/domain/lecture/ExerciseUnit.java:65]: Does it make sense to load the `learningGoals` if this happens - or is it fine that we just pretend there are none?

Comment (sleiss) [reply src/main/java/de/tum/in/www1/artemis/domain/lecture/ExerciseUnit.java:65]: I did not realize that this method is within the domain object so we can not load it here - If this turns out to be a problem in the future, we could move this logic to a service/repository.

File: src/main/java/de/tum/in/www1/artemis/web/rest/LearningGoalResource.java (hunk 5 above)
Comment (ge65cer) [reply src/main/java/de/tum/in/www1/artemis/web/rest/LearningGoalResource.java:200]: Fixed.

**Review by stefanwaldhauser — APPROVED** [review]
Approve! Every one of my comments addressed! Nice work!

**Pull request merged**